drltrace/drltrace.c

```diff
@@ -104,6 +105,119 @@ static app_pc exe_start;
+        if (dr_safe_read((void *)arg->value, arg->size, &deref, NULL))
+            dr_fprintf(outf, (leading_zeroes ? " => " PFX : " => " PIFX), deref);
+    }
+}
+
+static void
+print_arg(drsys_arg_t *arg)
+{
+    dr_fprintf(outf, "\n    arg %d: ", arg->ordinal);
+    switch (arg->type) {
+    case DRSYS_TYPE_VOID:         print_simple_value(arg, true); break;
+    case DRSYS_TYPE_POINTER:      print_simple_value(arg, true); break;
+    case DRSYS_TYPE_BOOL:         print_simple_value(arg, false); break;
+    case DRSYS_TYPE_INT:          print_simple_value(arg, false); break;
+    case DRSYS_TYPE_SIGNED_INT:   print_simple_value(arg, false); break;
+    case DRSYS_TYPE_UNSIGNED_INT: print_simple_value(arg, false); break;
+    case DRSYS_TYPE_HANDLE:       print_simple_value(arg, false); break;
+    case DRSYS_TYPE_NTSTATUS:     print_simple_value(arg, false); break;
+    case DRSYS_TYPE_ATOM:         print_simple_value(arg, false); break;
+    default: {
+        if (arg->value == 0)
+            dr_fprintf(outf, "<null>");
+        else
+            dr_fprintf(outf, PFX, arg->value);
+    }
+    }
+
+    dr_fprintf(outf, " (%s%s%stype=%s%s, size=" PIFX ")",
+              (arg->arg_name == NULL) ? "" : "name=",
+              (arg->arg_name == NULL) ? "" : arg->arg_name,
+              (arg->arg_name == NULL) ? "" : ", ",
+              (arg->type_name == NULL) ? "\"\"" : arg->type_name,
+              (arg->type_name == NULL ||
+              TESTANY(DRSYS_PARAM_INLINED|DRSYS_PARAM_RETVAL, arg->mode)) ? "" : "*",
+              arg->size);
+}
+
+static bool
+drlib_iter_arg_cb(drsys_arg_t *arg, void *wrapcxt)
+{
+    if (arg->ordinal == -1)
+        return true;
+    if (arg->ordinal >= options.max_args)
+        return false; /* limit number of arguments to be printed */
+
+    arg->value = (ptr_uint_t)drwrap_get_arg(wrapcxt, arg->ordinal);
+
+    print_arg(arg);
+    return true; /* keep going */
+}
+
+static void
+print_args_unknown_call(app_pc func, void *wrapcxt)
+{
+    uint i;
+    void *drcontext = drwrap_get_drcontext(wrapcxt);
+    dr_fprintf(outf, "(");
+    DR_TRY_EXCEPT(drcontext, {
+        for (i = 0; i < options.unknown_args; i++) {
+            dr_fprintf(outf, "%s" PFX, (i != 0) ? ", " : "",
+                       drwrap_get_arg(wrapcxt, i));
+        }
+    }, {
+        dr_fprintf(outf, "<invalid memory>");
+        /* Just keep going */
+    });
+    dr_fprintf(outf, ")");
+}
+
+static void
+print_symbolic_args(const char *name, void *wrapcxt, app_pc func)
+{
+    drmf_status_t res;
+    drsys_syscall_t *syscall;
+    char syscall_name[MAX_LIBCALL_NAME];
+
+    if (options.max_args > 0) {
+        if (name[0] == 'Z' && name[1] == 'w')
+            dr_snprintf(syscall_name, BUFFER_SIZE_ELEMENTS(syscall_name), "Nt%s", name+2);
+        else
+            dr_snprintf(syscall_name, BUFFER_SIZE_ELEMENTS(syscall_name), "%s", name);
+
+        res = drsys_name_to_syscall(syscall_name, &syscall);
```

Comment:
this is correct both versions work. We have an assert at https://github.com/DynamoRIO/drmemory/blob/master/drsyscall/drsyscall.c#L251